encode.go

```diff
@@ -11,74 +11,92 @@ import (
-	} else if data.Type == XT { // Table
-		binary.Write(dbuf, order, data.Attr)
+func write(buf *bytes.Buffer, order binary.ByteOrder, k *K) error {
+	binary.Write(buf, order, k.Type)
+	if K0 <= k.Type && k.Type <= KT { // All Lists
+		binary.Write(buf, order, k.Attr)
+		binary.Write(buf, order, int32(reflect.ValueOf(k.Data).Len()))
+	} else if k.Type == XT { // Table
+		binary.Write(buf, order, k.Attr)
 	}
 
-	switch data.Type {
+	switch k.Type {
 	case K0:
-		for _, toSend := range data.Data.([]*K) {
-			err := writeData(dbuf, order, toSend)
+		for _, toSend := range k.Data.([]*K) {
+			err := write(buf, order, toSend)
 			if err != nil {
 				return err
 			}
 		}
-	case -KB, -UU, -KG, -KH, -KI, -KJ, -KE, -KF, -KC, -KD, -KT, // TODO: case: -KM , -KZ, -KN, -KU, -KV
-		KB, UU, KG, KH, KI, KJ, KE, KF, KM, KD, KZ, KN, KU, KV, KT:
+	case -KB, -UU, -KG, -KH, -KI, -KJ, -KE, -KF, -KC, -KM , -KZ, -KN, -KU, -KV,
+		+KB, +UU, +KG, +KH, +KI, +KJ, +KE, +KF, +KM, +KZ, +KN, +KU, +KV:
 		// Scalar & Vector of boolean, int variants, float variant, and bytes.
-		// Note: UUID is backed by byte array of length 16
-		binary.Write(dbuf, order, data.Data)
+		// Note: UUID is back by byte array of length 16
+		binary.Write(buf, order, k.Data)
 	case KC: // String
-		dbuf.WriteString(data.Data.(string))
+		buf.WriteString(k.Data.(string))
 	case -KS: // Scalar symbol
-		dbuf.WriteString(data.Data.(string))
-		binary.Write(dbuf, order, byte(0)) // Null terminator
+		buf.WriteString(k.Data.(string))
+		binary.Write(buf, order, byte(0)) // Null terminator
 	case KS: // Vector symbol
-		for _, toSend := range data.Data.([]string) {
-			dbuf.WriteString(toSend)
-			binary.Write(dbuf, order, byte(0)) // Null terminator
+		for _, toSend := range k.Data.([]string) {
+			buf.WriteString(toSend)
+			binary.Write(buf, order, byte(0)) // Null terminator
 		}
 	case -KP: // Scalar timestamp
-		binary.Write(dbuf, order, data.Data.(time.Time).Sub(qEpoch))
+		binary.Write(buf, order, k.Data.(time.Time).Sub(qEpoch))
 	case KP: // Vector timestamp
-		for _, toSend := range data.Data.([]time.Time) {
-			binary.Write(dbuf, order, toSend.Sub(qEpoch))
+		for _, toSend := range k.Data.([]time.Time) {
+			binary.Write(buf, order, toSend.Sub(qEpoch))
+		}
+	case -KD: // Scalar date
+		date := k.Data.(time.Time)
+		days := date.Sub(qEpoch) / (time.Hour * 24)
+		binary.Write(buf, order, int32(days))
+	case KD: // Vector date
+		for _, date := range k.Data.([]time.Time) {
+			days := date.Sub(qEpoch) / (time.Hour * 24)
+			binary.Write(buf, order, int32(days))
+		}
+	case -KT: // Scalar time
+		t := k.Data.(time.Time)
+		nanos := t.Hour()*int(time.Hour) + t.Minute()*int(time.Minute) + t.Second()*int(time.Second) + t.Nanosecond()
+		binary.Write(buf, order, int32(nanos/int(time.Millisecond)))
+	case KT: // Vector time
+		for _, t := range k.Data.([]time.Time) {
+			nanos := t.Hour()*int(time.Hour) + t.Minute()*int(time.Minute) + t.Second()*int(time.Second) + t.Nanosecond()
```

Comment:
Is the untyped `int` safe here?

Reply:
Actually you are right. It wasn't safe. Updated to use `int64`